### PR TITLE
0010 - feat: add postmortem principles

### PR DESCRIPTION
**Pull request by maxneuvians**
This PR adds a RFC around Postmortem principles at CDS. It is inspired by @dj2 previous work and draws learning from other organisations such as Google and Atlassian.

### CODE REVIEW

File: rfcs/0010-postmortem-principles.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+## Motivation
+We need to align as an organisation on the value of postmortems - what they give us, why they are worth investing in, and what the consequences of ignoring outcomes are.
+
+## Proposal
+
+### People will make mistakes
+It is completely unreasonable to expect humans to never make a mistake, otherwise they would never learn. When people make mistakes, we assume they acted with the best intentions based on the information they had at the time. A postmortem needs to honestly and objectively examine the circumstances that led to the fault so we can find the true root cause(s) and mitigate them.
+
+### People need to feel psychologically safe
+Witnessing the public shaming and blaming of a co-worker for mistakes is a surefire way of reducing the confidence and productivity of others. Future incidences will most likely be covered up and the organisation as a whole will suffer from lost productivity and low morale.
```
Comment (patheard) [inline rfcs/0010-postmortem-principles.md:29]: You could add that covering up mistakes reinforces the blame cycle and leads to more incidents (the system becomes weaker overall).

File: rfcs/0010-postmortem-principles.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+---
+
+[La version française suit.](#résumé)
+
+# CDS Principles for running incident response and doing postmortems
+
+## Summary
+Things go wrong. Software breaks. Partnerships aren’t smooth. Delivery takes longer than expected. Sometimes, things go right. In all cases, we should take the time to look back on what happened, how it happened and what we can do better in the future.
+
+Postmortems are a tool for determining what happened, documenting the timelines and discovering what went right and what went wrong. They help us understand how a given outcome occurred and, if necessary, provide the action items to ensure that outcome doesn’t happen again in the future.
```
Comment (patheard) [inline rfcs/0010-postmortem-principles.md:18]: Could we talk about how postmortems, when adopted properly, strengthen the services delivered by that org?

File: rfcs/0010-postmortem-principles.md (hunk 1 above)
Comment (dj2) [reply rfcs/0010-postmortem-principles.md:29]: +1 that's a great point.

File: rfcs/0010-postmortem-principles.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+## Proposal
+
+### People will make mistakes
+It is completely unreasonable to expect humans to never make a mistake, otherwise they would never learn. When people make mistakes, we assume they acted with the best intentions based on the information they had at the time. A postmortem needs to honestly and objectively examine the circumstances that led to the fault so we can find the true root cause(s) and mitigate them.
+
+### People need to feel psychologically safe
+Witnessing the public shaming and blaming of a co-worker for mistakes is a surefire way of reducing the confidence and productivity of others. Future incidences will most likely be covered up and the organisation as a whole will suffer from lost productivity and low morale.
+
+### Incidents are an opportunity to learn and reflect
+In the broadest sense, an incident is any unplanned work you encounter in your day-to-day operations. Good judgement is required to know when unplanned events require an incident response, as seemingly innocuous events may have terrifying root causes. This will come with experience running incident responses, which means it is better to err on running more than running less.
```
Comment (patheard) [inline rfcs/0010-postmortem-principles.md:32]: Would it be useful to give some examples of what could trigger a postmortem?  Or would that be something for a future doc?  Google gives some good examples in their https://sre.google/sre-book/postmortem-culture/ post:

- User-visible downtime or degradation beyond a certain threshold
- Data loss of any kind
- On-call engineer intervention (release rollback, rerouting of traffic, etc.)
- A resolution time above some threshold
- A monitoring failure (which usually implies manual incident discovery)

File: rfcs/0010-postmortem-principles.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+### Timelines are important
+It is important to keep an up-to-date timeline of events during an incident as it will be the only moment in time when the information is fresh in your mind. Humans naturally collapse details as time goes by and important nuances will be lost. Similarly, it will give team members joining the incident later on a good overview of where the process is at.
+
+### Action items need ownership and accountability
+We can have the best postmortem process in the world, but have it amount to nothing unless we follow up on the action items identified and hold ourselves accountable to them. It is not enough to identify them, they need to be put into backlogs and actioned with dedicated owners.
```
Comment (dj2) [inline rfcs/0010-postmortem-principles.md:38]: Maybe s/need/MUST/ for putting into backlogs?

File: rfcs/0010-postmortem-principles.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+### Incidents and postmortems take time
+The creation, review and completion of postmortems should be given emphasis by the management team. It should be made clear that recording our past successes and failures is a priority. That there is time, and room, for a team to do their postmortem, do it correctly, and follow up as needed.
```
Comment (dj2) [inline rfcs/0010-postmortem-principles.md:41]: nit: do them correctly

File: rfcs/0010-postmortem-principles.md (hunk 4 above)
Comment (patheard) [inline rfcs/0010-postmortem-principles.md:38]: I also like the idea of reviewing postmortems with other teams to highlight how things can break and how they were fixed (even if it isn't directly related to the work "you" do).  It would help build psychological safety.

File: rfcs/0010-postmortem-principles.md (hunk 3 above)
Comment (CalvinRodo) [reply rfcs/0010-postmortem-principles.md:32]: This is definitely something we've struggled with and I think having clear guidelines around what constitutes an incident would be useful however we would need to make sure it's clear this list can never be definitive and incidents might fall outside of that list.

File: rfcs/0010-postmortem-principles.md (hunk 4 above)
Comment (dj2) [reply rfcs/0010-postmortem-principles.md:38]: I'd love to see this happen in the weekly dev community meeting. Take 5-10 minutes and run through a postmortem, what happened, how we caught it, what AIs were listed and how they were handled.

File: rfcs/0010-postmortem-principles.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+Things go wrong. Software breaks. Partnerships aren’t smooth. Delivery takes longer than expected. Sometimes, things go right. In all cases, we should take the time to look back on what happened, how it happened and what we can do better in the future.
+
+Postmortems are a tool for determining what happened, documenting the timelines and discovering what went right and what went wrong. They help us understand how a given outcome occurred and, if necessary, provide the action items to ensure that outcome doesn’t happen again in the future.
+
+## Motivation
+We need to align as an organisation on the value of postmortems - what they give us, why they are worth investing in, and what the consequences of ignoring outcomes are.
+
+## Proposal
+
+### People will make mistakes
```
Comment (CalvinRodo) [inline rfcs/0010-postmortem-principles.md:25]: Can we rephrase this title so it sounds like a proposal and less as a statement of fact. 

Maybe something like 

"We acknowledge that people will make mistakes"

File: rfcs/0010-postmortem-principles.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+It is completely unreasonable to expect humans to never make a mistake, otherwise they would never learn. When people make mistakes, we assume they acted with the best intentions based on the information they had at the time. A postmortem needs to honestly and objectively examine the circumstances that led to the fault so we can find the true root cause(s) and mitigate them.
+
+### People need to feel psychologically safe
```
Comment (CalvinRodo) [inline rfcs/0010-postmortem-principles.md:28]: Again can we rephrase this as a proposal, maybe like: 

"We will create an environment of psychological safety."

File: rfcs/0010-postmortem-principles.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+### Incidents are an opportunity to learn and reflect
+In the broadest sense, an incident is any unplanned work you encounter in your day-to-day operations. Good judgement is required to know when unplanned events require an incident response, as seemingly innocuous events may have terrifying root causes. This will come with experience running incident responses, which means it is better to err on running more than running less.
+
+### Timelines are important
+It is important to keep an up-to-date timeline of events during an incident as it will be the only moment in time when the information is fresh in your mind. Humans naturally collapse details as time goes by and important nuances will be lost. Similarly, it will give team members joining the incident later on a good overview of where the process is at.
+
+### Action items need ownership and accountability
+We can have the best postmortem process in the world, but have it amount to nothing unless we follow up on the action items identified and hold ourselves accountable to them. It is not enough to identify them, they need to be put into backlogs and actioned with dedicated owners.
+
+### Incidents and postmortems take time
```
Comment (CalvinRodo) [inline rfcs/0010-postmortem-principles.md:40]: Again proposal vs statement: 

"Time needs to be made for incidents and postmortems"

File: rfcs/0010-postmortem-principles.md [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+It is completely unreasonable to expect humans to never make a mistake, otherwise they would never learn. When people make mistakes, we assume they acted with the best intentions based on the information they had at the time. A postmortem needs to honestly and objectively examine the circumstances that led to the fault so we can find the true root cause(s) and mitigate them.
+
+### People need to feel psychologically safe
+Witnessing the public shaming and blaming of a co-worker for mistakes is a surefire way of reducing the confidence and productivity of others. Future incidences will most likely be covered up and the organisation as a whole will suffer from lost productivity and low morale.
+
+### Incidents are an opportunity to learn and reflect
+In the broadest sense, an incident is any unplanned work you encounter in your day-to-day operations. Good judgement is required to know when unplanned events require an incident response, as seemingly innocuous events may have terrifying root causes. This will come with experience running incident responses, which means it is better to err on running more than running less.
+
+### Timelines are important
```
Comment (CalvinRodo) [inline rfcs/0010-postmortem-principles.md:34]: Again statement vs proposal, I don't know how to rephrase this off the top of my head.

**Review by CalvinRodo — COMMENTED** [review]
Should we link to our incident management process in this RFC? 

Do we need to outline the responsibilities of the Incident Commander in regards to an Incident or should that be covered in another RFC.

File: rfcs/0010-postmortem-principles.md (hunk 1 above)
Comment (maxneuvians) [reply rfcs/0010-postmortem-principles.md]: Thank you!

File: rfcs/0010-postmortem-principles.md (hunk 2 above)
Comment (maxneuvians) [reply rfcs/0010-postmortem-principles.md]: Thank you!

File: rfcs/0010-postmortem-principles.md (hunk 3 above)
Comment (maxneuvians) [reply rfcs/0010-postmortem-principles.md]: Thank you!

File: rfcs/0010-postmortem-principles.md (hunk 5 above)
Comment (maxneuvians) [reply rfcs/0010-postmortem-principles.md]: Thank you!

File: rfcs/0010-postmortem-principles.md (hunk 4 above)
Comment (maxneuvians) [reply rfcs/0010-postmortem-principles.md]: Thank you!

File: rfcs/0010-postmortem-principles.md (hunk 7 above)
Comment (maxneuvians) [reply rfcs/0010-postmortem-principles.md]: Thank you!

File: rfcs/0010-postmortem-principles.md (hunk 9 above)
Comment (maxneuvians) [reply rfcs/0010-postmortem-principles.md]: Thank you!

File: rfcs/0010-postmortem-principles.md (hunk 6 above)
Comment (maxneuvians) [reply rfcs/0010-postmortem-principles.md]: Thank you!

**Comment by maxneuvians** [comment]
Re: Should we link to our incident management process in this RFC?

Should be a new RFC

File: rfcs/0010-postmortem-principles.md (hunk 4 above)
Comment (maxneuvians) [reply rfcs/0010-postmortem-principles.md]: Thank you!

**Comment by dj2** [comment]
I believe this will fix #13 when committed.

**Comment by dj2** [comment]
Can you sync and run the formatter?

File: development/rfcs/0010-postmortem-principles.md [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+Witnessing the public shaming and blaming of a co-worker for mistakes is a surefire way of reducing the confidence and productivity of others. Future incidences will most likely be covered up and reinforce the blame cycle, leading to more incidents, making the system weaker overall.
+
+### Incidents are an opportunity to learn and reflect
+
+In the broadest sense, an incident is any unplanned work you encounter in your day-to-day operations. Good judgement is required to know when unplanned events require an incident response, as seemingly innocuous events may have terrifying root causes. This will come with experience running incident responses, which means it is better to err on running more than running less. Some examples of events that may trigger the incident process are:
+
+- User-visible downtime or degradation beyond a certain threshold
+- Data loss of any kind
+- On-call engineer intervention (release rollback, rerouting of traffic, etc.)
+- A resolution time above some threshold
```
Comment (wmoussa-cds) [inline development/rfcs/0010-postmortem-principles.md:42]: do you mean a response time? (as in API response time)

File: development/rfcs/0010-postmortem-principles.md [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+### Time needs to be made for incidents and postmortems
+
+The creation, review and completion of postmortems should be given emphasis by the management team. It should be made clear that recording our past successes and failures is a priority. That there is time, and room, for a team to do their postmortem, do them correctly, and follow up as needed.
+
+## Drawbacks
+
+We assume that this is the best way for us to learn from past mistakes and successes. However, it requires significant investment of time and resources to do properly, which comes at the opportunity cost of other work. Should we not properly follow through on our commitments we may end up with a broken process that has wasted time and resources with no usable outcomes.
```
Comment (wmoussa-cds) [inline development/rfcs/0010-postmortem-principles.md:59]: I think 'significant investment of time' might sound too much.I would remove _significant_

File: development/rfcs/0010-postmortem-principles.md [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+
+## Alternatives
+
+Time invested into postmortems could be invested into other work, while incidents are centrally monitored and managed.
```
Comment (wmoussa-cds) [inline development/rfcs/0010-postmortem-principles.md:63]: I don't understand this alternative. Possible to expand?

One alternative (not ideal of course): only document incident with long downtime or damage to the users experience

File: development/rfcs/0010-postmortem-principles.md [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+## Open Questions
+
+How do these align with the principles of our organisation?
```
Comment (wmoussa-cds) [inline development/rfcs/0010-postmortem-principles.md:67]: I will add: How we make sure time is allocated properly for this process

**Review by wmoussa-gc — APPROVED** [review]

File: development/rfcs/0010-postmortem-principles.md (hunk 10 above)
Comment (maxneuvians) [reply development/rfcs/0010-postmortem-principles.md]: Clarified - thank you

File: development/rfcs/0010-postmortem-principles.md (hunk 11 above)
Comment (maxneuvians) [reply development/rfcs/0010-postmortem-principles.md]: Thank you!

File: development/rfcs/0010-postmortem-principles.md (hunk 12 above)
Comment (maxneuvians) [reply development/rfcs/0010-postmortem-principles.md]: Thank you!

File: development/rfcs/0010-postmortem-principles.md (hunk 13 above)
Comment (maxneuvians) [reply development/rfcs/0010-postmortem-principles.md:68]: Added - thank you!
